media_cdn/dualtoken.rb

```diff
@@ -0,0 +1,179 @@
+    #     ip_ranges: a list of comma-separated IP ranges. Both IPv4 and IPv6 ranges are acceptable.
+    #                 For example: '203.0.113.0/24,2001:db8:4a7f:a732/64'
+    # Returns:
+    #     The Signed URL appended with the query parameters based on the
+    #     specified URL prefix and configuration.
+
+    decoded_key = Base64.urlsafe_decode64 base64_key
+    algo = signature_algorithm.downcase
+
+    # For most fields, the value we put in the token and the value we must sign
+    # are the same.  The FullPath and Headers use a different string for the
+    # value to be signed compared to the token.  To illustrate this difference,
+    # we'll keep the token and the value to be signed separate.
+    tokens = []
+    to_sign = []
+
+    # check for `full_path` or `path_globs` or `url_prefix`
+    if full_path
+      tokens.append "FullPath"
+      to_sign.append "FullPath=#{full_path}"
+    elsif path_globs
+      path_globs = path_globs.strip
+      field = "PathGlobs=#{path_globs}"
+      tokens.append field
+      to_sign.append field
+    elsif url_prefix
+      encoded_url_prefix = base64_encoder(url_prefix).encode("utf-8")
+      field = "URLPrefix=#{encoded_url_prefix}"
+      tokens.append field
+      to_sign.append field
+    else
+      raise "User input missing: one of `url_prefix`, `full_path`, or `path_globs` must be specified."
+    end
+
+    # check & parse optional params
+    # start_time
+    if start_time
+      start_time_utc = start_time.utc.to_i
+
+      field = "Starts=#{start_time_utc}"
+      tokens.append field
+      to_sign.append field
+    end
+
+    # expiration_time
+    expiration_utc = if expiration_time
+                       expiration_time.utc.to_i
+                     else
+                       Time.now.to_i + 300
+                     end
+    field = "Expires=#{expiration_utc}"
+    tokens.append field
+    to_sign.append field
+
+    # session_id
+    if session_id
+      field = "SessionID=#{session_id}"
+      tokens.append field
+      to_sign.append field
+    end
+
+    # data
+    if data
+      field = "Data=#{data}"
+      tokens.append field
+      to_sign.append field
+    end
+
+    # headers
+    if headers
+      header_names = []
+      header_pairs = []
+      # for each in headers
+      headers.each do |header|
+        header_names.append header[:name]
+        header_pairs.append "#{header[:name]}=#{header[:value]}"
+      end
+      header_names = header_names.join ","
+      header_pairs = header_pairs.join ","
+      tokens.append "Headers=#{header_names}"
+      to_sign.append "Headers=#{header_pairs}"
+    end
+
+    # ip-range
+    if ip_ranges
+      encoded_ip_ranges = base64_encoder(ip_ranges.encode("ascii"))
+      field = "IPRanges=#{encoded_ip_ranges}"
+      tokens.append field
+      to_sign.append field
+    end
+
+    # generating token
+    to_sign = to_sign.join "~"
+    to_sign_bytes = to_sign.encode "utf-8"
```

Comment:
I'd write as `to_sign_bytes = to_sign.join('~').encode('utf-8')`so it doesn't change the type of `to_sign`and repurpose it.